pkg/imgpkg/image/keychain_test.go

```diff
@@ -0,0 +1,136 @@
+	assert.NoError(t, err)
+
+	assert.Equal(t, authn.Anonymous, auth)
+}
+
+func TestEnvAuthWhenEnvVarsProvided(t *testing.T) {
+	setTestEnv(t, "USERNAME", "user")
+	setTestEnv(t, "PASSWORD", "pass")
+	setTestEnv(t, "HOSTNAME", "localhost:9999")
+
+	defer unsetTestEnv(t)
+
+	keychain := Keychain("", "", "", false)
+	resource, err := name.NewRepository("localhost:9999/imgpkg_test")
+	assert.NoError(t, err)
+
+	auth, err := keychain.Resolve(resource)
+	assert.NoError(t, err)
+
+	assert.Equal(t, authn.FromConfig(authn.AuthConfig{
+		Username: "user",
+		Password: "pass",
+	}), auth)
+}
+
+func TestEnvAuthWhenEnvVarsProvidedWithMultipleRegistries(t *testing.T) {
+	setTestEnv(t, "USERNAME_0", "user_0")
+	setTestEnv(t, "PASSWORD_0", "pass_0")
+	setTestEnv(t, "HOSTNAME_0", "localhost:0000")
+
+	setTestEnv(t, "USERNAME_1", "user_1")
+	setTestEnv(t, "PASSWORD_1", "pass_1")
+	setTestEnv(t, "HOSTNAME_1", "localhost:1111")
+
+	defer unsetTestEnv(t)
+
+	keychain := Keychain("", "", "", false)
+	resource, err := name.NewRepository("localhost:1111/imgpkg_test")
+	assert.NoError(t, err)
+
+	auth, err := keychain.Resolve(resource)
+	assert.NoError(t, err)
+
+	assert.Equal(t, authn.FromConfig(authn.AuthConfig{
+		Username: "user_1",
+		Password: "pass_1",
+	}), auth)
+}
+
+func TestRegistryToken(t *testing.T) {
+	setTestEnv(t, "HOSTNAME", "localhost:1111")
+	setTestEnv(t, "REGISTRY_TOKEN", "TOKEN")
+
+	defer unsetTestEnv(t)
+
+	keychain := Keychain("", "", "", false)
+	resource, err := name.NewRepository("localhost:1111/imgpkg_test")
+	assert.NoError(t, err)
+
+	auth, err := keychain.Resolve(resource)
+	assert.NoError(t, err)
+
+	assert.Equal(t, authn.FromConfig(authn.AuthConfig{
+		RegistryToken: "TOKEN",
+	}), auth)
+}
+
+func TestIdentityToken(t *testing.T) {
+	setTestEnv(t, "HOSTNAME", "localhost:1111")
+	setTestEnv(t, "IDENTITY_TOKEN", "ID_TOKEN")
+
+	defer unsetTestEnv(t)
+
+	keychain := Keychain("", "", "", false)
+	resource, err := name.NewRepository("localhost:1111/imgpkg_test")
+	assert.NoError(t, err)
+
+	auth, err := keychain.Resolve(resource)
+	assert.NoError(t, err)
+
+	assert.Equal(t, authn.FromConfig(authn.AuthConfig{
+		IdentityToken: "ID_TOKEN",
+	}), auth)
+}
+
+func TestAuthFromFlagHasPrecedenceOverEnvVars(t *testing.T) {
+	setTestEnv(t, "USERNAME", "envUser")
+	setTestEnv(t, "PASSWORD", "envPass")
+	setTestEnv(t, "HOSTNAME", "localhost:9999")
+
+	defer unsetTestEnv(t)
+
+	keychain := Keychain("flagUser", "flagPass", "", false)
+	resource, err := name.NewRepository("localhost:9999/imgpkg_test")
+	assert.NoError(t, err)
+
+	auth, err := keychain.Resolve(resource)
+	assert.NoError(t, err)
+
+	assert.Equal(t, &authn.Basic{
+		Username: "flagUser",
+		Password: "flagPass",
+	}, auth)
+}
+
+func setTestEnv(t *testing.T, key string, value string) {
+	assert.NoError(t, os.Setenv(globalPrefix+"_"+key, value))
+}
+
+func unsetTestEnv(t *testing.T) {
+	for _, env := range os.Environ() {
+		if strings.HasPrefix(env, globalPrefix) {
```

Comment:
```suggestion
		if strings.HasPrefix(env, "IMGPKG_REGISTRY_") {
```
For testing purposes, I prefer to have hardcoded the constant strings. This allows us to catch errors if for some reason these constants change or are incorrect.